### ISSUE TITLE
Enhance return type hints and param comment annotations

### ISSUE THREAD

**Issue by peter279k** [ar2labs/wiring#12]
As title, some methods inside classes and interfaces don't have the correct parameter type hints, return type hints, `@param` type comment annotations and `@return` type comment annotations.

It should modify them and it's easy for developers to know about the method usages via correct types.

**Comment by arcostasi** [comment]
Very good Peter, I will pull your improvements . Thank you

**Comment by peter279k** [comment]
@arcostasi, thanks for your reply.

I also try integrating the static analysis tool to help us to evaluate type hints and class relationships.

The analysis tool lists are as follows:

- phpstan
- phan
- psalm

I've ever used the `phpstan` before.

Do you prefer the static code analysis tools?

**Comment by arcostasi** [comment]
Hey Peter, all checks have failed in Travis: https://github.com/ar2labs/wiring/pull/13

**Comment by arcostasi** [comment]
Ok, no problem @peter279k, let's use the best tools! ;)

**Comment by arcostasi** [comment]
> @arcostasi, thanks for your reply.
> I also try integrating the static analysis tool to help us to evaluate type hints and class relationships.
> The analysis tool lists are as follows:
> 
> phpstan
> phan
> psalm
> 
> I've ever used the phpstan before.
> Do you prefer the static code analysis tools?

Ok, great @peter279k, let's use the best tools! ;)

**Comment by arcostasi** [comment]
Any pending issues with `phpstan` @peter279k ?

**Comment by peter279k** [comment]
@arcostasi,, thanks for your reply.

It seems that this [issue ](https://github.com/ar2labs/wiring/issues/14)is resolved and I keep this issue going to the next work.

**Comment by peter279k** [comment]
@arcostasi, since the PR #21 is created, most of problematic codes are resolved on `phpstan` `level 2`.

But some codes still have been problematic because of the class implementation approach.

The error messages are as follows and they're not fixed on PR #21.

```
 ------ ------------------------------------------------------------------------ 
  Line   Application.php                                                         
 ------ ------------------------------------------------------------------------ 
  17     PHPDoc tag @throws with type Wiring\Http\Exception\ErrorHandler is not  
         subtype of Throwable                                                    
  26     PHPDoc tag @throws with type Wiring\Http\Exception\ErrorHandler is not  
         subtype of Throwable                                                    
 ------ ------------------------------------------------------------------------ 

 ------ ------------------------------------------------------- 
  Line   Http/Controller/AbstractJsonController.php             
 ------ ------------------------------------------------------- 
  53     Call to an undefined method                            
         Psr\Http\Message\ResponseFactoryInterface::getBody().  
 ------ ------------------------------------------------------- 

 ------ ------------------------------------------------------------------------ 
  Line   Http/RequestHandler.php                                                 
 ------ ------------------------------------------------------------------------ 
  91     PHPDoc tag @throws with type Wiring\Http\Exception\ErrorHandler is not  
         subtype of Throwable                                                    
  107    PHPDoc tag @throws with type Wiring\Http\Exception\ErrorHandler is not  
         subtype of Throwable                                                    
  206    PHPDoc tag @param has invalid value ($router): Unexpected token         
         "$router", expected TOKEN_IDENTIFIER at offset 55                       
  322    PHPDoc tag @throws with type Wiring\Http\Exception\ErrorHandler is not  
         subtype of Throwable                                                    
 ------ ------------------------------------------------------------------------ 

 [ERROR] Found 7 errors                                                         

```

As you can see, these classes have the two problems.

- The `ErrorHandler` class will not throw the current Exception and it doesn't extends the `Throwable`. That's why the message is `it's not subtype of Throwable class`.
- Another problem is about undefined method inside `AbstractJsonController` class.
The `Psr\Http\Message\ResponseFactoryInterface::getBody(). ` need to be resolved.

**Comment by arcostasi** [comment]
> @arcostasi, since the PR #21 is created, most of problematic codes are resolved on phpstan level 2.
> But some codes still have been problematic because of the class implementation approach.
> The error messages are as follows and they're not fixed on PR #21.
>  ------ ------------------------------------------------------------------------ 
>   Line   Application.php                                                         
>  ------ ------------------------------------------------------------------------ 
>   17     PHPDoc tag @throws with type Wiring\Http\Exception\ErrorHandler is not  
>          subtype of Throwable                                                    
>   26     PHPDoc tag @throws with type Wiring\Http\Exception\ErrorHandler is not  
>          subtype of Throwable                                                    
>  ------ ------------------------------------------------------------------------ 
> 
>  ------ ------------------------------------------------------- 
>   Line   Http/Controller/AbstractJsonController.php             
>  ------ ------------------------------------------------------- 
>   53     Call to an undefined method                            
>          Psr\Http\Message\ResponseFactoryInterface::getBody().  
>  ------ ------------------------------------------------------- 
> 
>  ------ ------------------------------------------------------------------------ 
>   Line   Http/RequestHandler.php                                                 
>  ------ ------------------------------------------------------------------------ 
>   91     PHPDoc tag @throws with type Wiring\Http\Exception\ErrorHandler is not  
>          subtype of Throwable                                                    
>   107    PHPDoc tag @throws with type Wiring\Http\Exception\ErrorHandler is not  
>          subtype of Throwable                                                    
>   206    PHPDoc tag @param has invalid value ($router): Unexpected token         
>          "$router", expected TOKEN_IDENTIFIER at offset 55                       
>   322    PHPDoc tag @throws with type Wiring\Http\Exception\ErrorHandler is not  
>          subtype of Throwable                                                    
>  ------ ------------------------------------------------------------------------ 
> 
>  [ERROR] Found 7 errors                                                         
> 
> 
> As you can see, these classes have the two problems.
> 
> The ErrorHandler class will not throw the current Exception and it doesn't extend the Throwable.
> It seems that the ErrorHandler class only returns the Exception class directly.
> That's why the message is it's not subtype of Throwable class.
> Another problem is about undefined method call inside AbstractJsonController class.
> The Psr\Http\Message\ResponseFactoryInterface::getBody().  error message needs to be resolved.

Thanks @peter279k , I'll be working on this fixes. I appreciate your considerations! Congratulations!

**Comment by arcostasi** [comment]
@peter279k, @vinyvicente, The levels 2, 3 and 4 has been fixed in [Resolves issue #12 via phpstan level 3 and 4](https://github.com/ar2labs/wiring/commit/297da915b8e3c5274b502d886c3d946693ab4f9c), we need to fix the test errors in level 5 for continue the next levels 6 and 7.

The error messages are as follows and they're not fixed on level 5.

```
------ ---------------------------------------------------------------------------------------------------------------- 
  Line   tests/ApplicationTest.php                                                                                       
 ------ ---------------------------------------------------------------------------------------------------------------- 
  21     Parameter #1 $container of class Wiring\Application constructor expects Psr\Container\ContainerInterface,       
         PHPUnit\Framework\MockObject\MockObject given.                                                                  
  21     Parameter #2 $request of class Wiring\Application constructor expects Psr\Http\Message\ServerRequestInterface,  
         PHPUnit\Framework\MockObject\MockObject given.                                                                  
  21     Parameter #3 $response of class Wiring\Application constructor expects Psr\Http\Message\ResponseInterface,      
         PHPUnit\Framework\MockObject\MockObject given.                                                                  
 ------ ---------------------------------------------------------------------------------------------------------------- 

 ------ -------------------------------------------------------------------------------------------------------- 
  Line   tests/Http/Controller/ControllerTest.php                                                                
 ------ -------------------------------------------------------------------------------------------------------- 
  35     Parameter #1 $container of class Wiring\Tests\Http\Controller\SimpleMockController constructor expects  
         Psr\Container\ContainerInterface, PHPUnit\Framework\MockObject\MockObject given.                        
  35     Parameter #2 $response of class Wiring\Tests\Http\Controller\SimpleMockController constructor expects   
         Psr\Http\Message\ResponseInterface, PHPUnit\Framework\MockObject\MockObject given.                      
  60     Parameter #1 $container of class Wiring\Tests\Http\Controller\SimpleJsonController constructor expects  
         Psr\Container\ContainerInterface, PHPUnit\Framework\MockObject\MockObject given.                        
  60     Parameter #2 $response of class Wiring\Tests\Http\Controller\SimpleJsonController constructor expects   
         Psr\Http\Message\ResponseInterface, PHPUnit\Framework\MockObject\MockObject given.                      
 ------ -------------------------------------------------------------------------------------------------------- 

 ------ --------------------------------------------------------------------------------------------------- 
  Line   tests/Http/Exception/ExceptionsTest.php                                                            
 ------ --------------------------------------------------------------------------------------------------- 
  52     Parameter #1 $response of method Wiring\Http\Exception\HttpException::buildJsonResponse() expects  
         Psr\Http\Message\ResponseInterface, PHPUnit\Framework\MockObject\MockObject given.                 
  53     Parameter #1 $response of method Wiring\Http\Exception\HttpException::buildResponse() expects      
         Psr\Http\Message\ResponseInterface, PHPUnit\Framework\MockObject\MockObject given.                 
 ------ --------------------------------------------------------------------------------------------------- 

 [ERROR] Found 9 errors     
```

**Comment by peter279k** [comment]
@arcostasi, thanks for your reply.

I will check that level 5,6 and 7.

**Comment by peter279k** [comment]
@arcostasi, since the PR #22 is created, it has been resolved many issues about following command:

```
php vendor/bin/phpstan analyse src/ --level=7 --no-progress
```

The `phpstan` level 5 and level 6 don't have any error on `src` folder.

But it still has some issues that you need consider and they're not resolved by PR #22.

The analysis error messages are as follows:

```
 ------ ------------------------------------------------------------------------- 
  Line   Http/Exception/ErrorHandler.php                                          
 ------ ------------------------------------------------------------------------- 
  71     Property Wiring\Http\Exception\ErrorHandler::$logger                     
         (Psr\Log\LoggerInterface) does not accept Psr\Log\LoggerInterface|null.  
 ------ ------------------------------------------------------------------------- 

 ------ ----------------------------------------------------------------------- 
  Line   Http/Exception/HttpException.php                                       
 ------ ----------------------------------------------------------------------- 
  106    Cannot call method getMessage() on Exception|null.                     
  109    Parameter #1 $object of function method_exists expects object|string,  
         Exception|null given.                                                  
  147    Cannot call method getMessage() on Exception|null.                     
 ------ ----------------------------------------------------------------------- 

 ------ ------------------------------------------ 
  Line   Strategy/JsonStrategy.php                 
 ------ ------------------------------------------ 
  74     Cannot call method getBody() on           
         Psr\Http\Message\ResponseInterface|null.  
  78     Cannot call method getBody() on           
         Psr\Http\Message\ResponseInterface|null.  
  83     Cannot call method withStatus() on        
         Psr\Http\Message\ResponseInterface|null.  
 ------ ------------------------------------------ 

 ------ ------------------------------------------------------------------------------------------------ 
  Line   Traits/ContainerAwareTrait.php (in context of class Wiring\Http\Controller\AbstractController)  
 ------ ------------------------------------------------------------------------------------------------ 
  73     Cannot call method has() on Psr\Container\ContainerInterface|null.                              
  90     Parameter #1 $object of function method_exists expects object|string,                           
         Psr\Container\ContainerInterface|null given.                                                    
  112    Parameter #1 $object of function method_exists expects object|string,                           
         Psr\Container\ContainerInterface|null given.                                                    
  135    Parameter #1 $object of function method_exists expects object|string,                           
         Psr\Container\ContainerInterface|null given.                                                    
 ------ ------------------------------------------------------------------------------------------------ 

 [ERROR] Found 11 errors                                                        

```

They're about related `Psr` interface type compatibility.

I think they're resolved by @arcostasi because they're about the core issues.

**Comment by peter279k** [comment]
BTW, the `phpstan` is for `tests` classes I think is another issue and we should focus on classes in `src` folder at this moment.

**Comment by peter279k** [comment]
@arcostasi, good work!.

Now I will use `phpstan` tool to do static code analysis for `tests` folder :).

**Comment by peter279k** [comment]
@arcostasi, I've create the PR #23.

Once this PR is merged, the `phpstan` level `6` and `level `7` don't have no errors.

![image](https://user-images.githubusercontent.com/9021747/64473668-2a856780-d19c-11e9-9683-b61d0090d6f4.png)

Once the PR is merged, we can add the `phpstan analyse src tests --level=max -c phpstan.neon --no-progress` command on `script` setting during Travis CI build.

**Comment by arcostasi** [comment]
> @arcostasi, I've created the PR #23.
> Once this PR is merged, the phpstan level 6 and level 7 have no errors.

Great work @peter279k, congratulations! ;)

> Once the PR is merged, we can add the phpstan analyse src tests --level=max -c phpstan.neon --no-progress command on script setting during Travis CI build.

Ok, already been merged! let's add... thank you!

**Comment by peter279k** [comment]
It seems that this issue has been accomplished.

Closed.

**Comment by AnjaLiebermann** [comment]
Do I see this correctly, that the solution ist to not check test classes with phpstan? 

I use $ bin/phpstan -V
PHPStan - PHP Static Analysis Tool 0.12.38

But have the error mentioned by arcostasi  in level 5 in my tests. Of course if I decide to not check my unittest classes I am fine ;)

**Comment by peter279k** [comment]
Thanks for your reply. We use the `PHP Static Analysis Tool 0.12.39` and using the `phpstan analyse src tests --level=max --no-progress --ansi` and it doesn't report any error :).

You can check the `scripts` block on `composer.json`.

**Comment by AnjaLiebermann** [comment]
@peter279k with that setting I get errors.

**Comment by AnjaLiebermann** [comment]
@peter279k  with that seting I get exactly those errors ( and of course even more, no surprise) 

Output:
`Parameter #3 $detectionService of class App\Domain\ChargeService constructor expects App\Domain\DetectionServiceInterface,
         PHPUnit\Framework\MockObject\MockObject given.`

The code causing it: 

`

    private function mockDetectionService(): \PHPUnit\Framework\MockObject\MockObject
    {
        /* @var DetectionServiceInterface detailService */
        $detectionService = $this
            ->getMockBuilder(DetectionService::class)
            ->disableOriginalConstructor()
            ->getMock();
        $detectionService
            ->method('create')
            ->willReturn($this->setDetection(false));
        $detectionService
            ->method('update')
            ->willReturn($this->setDetection(true));

        return $detectionService;
    }`

This $detectionServce is the used as a mock: 

`$this->chargeService = new ChargeService($this->em, $detailService, $detectionService);`

**Comment by peter279k** [comment]
You can refer this [PR](https://github.com/ar2labs/wiring/pull/23) or consider using the [phpstan-phpunit](https://github.com/phpstan/phpstan-phpunit) and it can help you to pass level 5 with PHPStan static code analysis work :).

**Comment by arcostasi** [comment]
Hello @AnjaLiebermann, here also passed the tests with phpstan 0.12.39 as mentioned by @peter279k

I'm using the following steps:

`composer phpstan`

output:
> phpstan analyse src tests --level=max --no-progress --ansi
Note: Using configuration file .\phpstan.neon
> 
> [OK] No errors

PHP 7.4.8
composer 1.10.10
phpstan 0.12.39

My phpstan are installed globally:

`composer global require phpstan/phpstan`

these analyzes are verified in travis according to the [latest Build #168](https://travis-ci.org/github/ar2labs/wiring/jobs/711750483), on line **372**

Please, if you have any questions or problems let us know. thankful

**Comment by AnjaLiebermann** [comment]
@peter279k Thank you for your prompt reply! I installed  phpstan-phpunit  and had a look at the mentioned PR, but I am not sure what I have to change in my test code to make it pass phpstan. 

"To fix createMock method issue, using the createMethodNameMock method to return mocked class instead." 
Where do I use it, do I have to use it? That is not really clear to me. 
Do I replace createMock with  createMethodNameMock in my code? What do I have to extend / include to make that available? 
Is there an example anywhere?

**Comment by AnjaLiebermann** [comment]
@peter279k thank you for your answer, but creating stub files to make my unittests pass phpstan is too much overhead to me. (If I understood this solution correctly?)  

I use phpstan, I write unit and functional tests, which are useful but take up a whole lot of time, so now writing extra code just to let my test pass phpstan? No thanks.... I guess I just exclude them from the check and am done and won't get more weird looks from my colleagues and remarks like  "Do you think level 5 is really necessary?"

**Comment by peter279k** [comment]
> Do you think level 5 is really necessary?

Well, it's up to you, the static code analysis checking usually is not for code errors.

And you can also set ignoring this PHPUnit mock codes checking on `phpstan.neon` file.

**Comment by AnjaLiebermann** [comment]
@peter279k actually the latter is a good advice.
